File: eco-api/src/main/java/com/willfp/eco/core/entities/ai/TargetGoals.java
package com.willfp.eco.core.entities.ai;

import com.google.common.collect.HashBiMap;
import com.willfp.eco.core.config.interfaces.Config;
import com.willfp.eco.core.entities.ai.target.TargetGoalDefendVillage;
import com.willfp.eco.core.entities.ai.target.TargetGoalHurtBy;
import com.willfp.eco.core.entities.ai.target.TargetGoalNearestAttackable;
import com.willfp.eco.core.entities.ai.target.TargetGoalNearestAttackableWitch;
import com.willfp.eco.core.entities.ai.target.TargetGoalNearestHealableRaider;
import com.willfp.eco.core.entities.ai.target.TargetGoalNonTameRandom;
import com.willfp.eco.core.entities.ai.target.TargetGoalOwnerHurtBy;
import com.willfp.eco.core.entities.ai.target.TargetGoalOwnerTarget;
import com.willfp.eco.core.entities.ai.target.TargetGoalResetUniversalAnger;
import com.willfp.eco.core.serialization.KeyedDeserializer;
import org.bukkit.NamespacedKey;
import org.bukkit.entity.Mob;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Map;

/**
 * Class to manage target goals.
 */
public final class TargetGoals {
    /**
     * All registered deserializers.
     */
    private static final Map<NamespacedKey, KeyedDeserializer<? extends TargetGoal<?>>> BY_KEY = HashBiMap.create();

    static {
        register(TargetGoalDefendVillage.DESERIALIZER);
        register(TargetGoalHurtBy.DESERIALIZER);
        register(TargetGoalNearestAttackable.DESERIALIZER);
        register(TargetGoalNearestAttackableWitch.DESERIALIZER);
        register(TargetGoalNearestHealableRaider.DESERIALIZER);
        register(TargetGoalNonTameRandom.DESERIALIZER);
        register(TargetGoalOwnerTarget.DESERIALIZER);
        register(TargetGoalOwnerHurtBy.DESERIALIZER);
        register(TargetGoalResetUniversalAnger.DESERIALIZER);
    }

    /**
     * Get deserializer by key.
     *
     * @param key The key.
     * @return The deserializer, or null if not found.
     */
    @Nullable
    public static KeyedDeserializer<? extends TargetGoal<? extends Mob>> getByKey(@NotNull final NamespacedKey key) {
        return BY_KEY.get(key);
    }

    /**
     * Get deserializer by key, with a defined type (to prevent cluttering code with unsafe casts).
     *
     * @param key   The key.
     * @param clazz The type of target goal.
     * @param <T>   The type of mob the goal can be applied to.
     * @return The deserializer, or null if not found.
     */
    @Nullable
    @SuppressWarnings({"unchecked", "unused"})
    public static <T extends Mob> KeyedDeserializer<TargetGoal<T>> getByKeyOfType(@NotNull final NamespacedKey key,
                                                                                  @NotNull final Class<T> clazz) {
        return (KeyedDeserializer<TargetGoal<T>>) BY_KEY.get(key);
    }

    /**
     * Apply goal to entity given key and config.
     * <p>
     * If the key or config are invalid, the goal will not be applied.
     *
     * @param entity   The entity.
     * @param key      The key.
     * @param config   The config.
     * @param priority The priority.
     * @param <T>      The entity type.
     * @return The entity.
     */
    @NotNull
    @SuppressWarnings("unchecked")
    public static <T extends Mob> T applyToEntity(@NotNull final T entity,
                                                  @NotNull final NamespacedKey key,
                                                  @NotNull final Config config,
                                                  final int priority) {
        KeyedDeserializer<TargetGoal<T>> deserializer = getByKeyOfType(key, (Class<T>) entity.getClass());
        if (deserializer == null) {
            return entity;
        }

        TargetGoal<T> goal = deserializer.deserialize(config);

        if (goal == null) {
            return entity;
        }

        return goal.addToEntity(entity, priority);
    }

    /**
     * Register a deserializer for a target goal.
     *
     * @param toRegister The target goal to register.
     * @param <T>        The type of deserializer.
     * @return The deserializer.
     */
    @NotNull
    public static <T extends KeyedDeserializer<? extends TargetGoal<?>>> T register(@NotNull final T toRegister) {
        BY_KEY.put(toRegister.getKey(), toRegister);
        return toRegister;
    }

    private TargetGoals() {
        throw new UnsupportedOperationException("This is a utility class and cannot be instantiated");
    }
}

File: eco-core/core-nms/common/src/main/kotlin/com/willfp/eco/internal/spigot/proxy/common/ai/TargetGoals.kt
package com.willfp.eco.internal.spigot.proxy.common.ai

import com.willfp.eco.core.entities.ai.CustomGoal
import com.willfp.eco.core.entities.ai.TargetGoal
import com.willfp.eco.core.entities.ai.target.TargetGoalHurtBy
import com.willfp.eco.core.entities.ai.target.TargetGoalNearestAttackable
import com.willfp.eco.core.entities.ai.target.TargetGoalNearestAttackableWitch
import com.willfp.eco.core.entities.ai.target.TargetGoalNearestHealableRaider
import com.willfp.eco.core.entities.ai.target.TargetGoalNonTameRandom
import com.willfp.eco.core.entities.ai.target.TargetGoalOwnerHurtBy
import com.willfp.eco.core.entities.ai.target.TargetGoalOwnerTarget
import com.willfp.eco.internal.spigot.proxy.common.ai.target.HurtByGoalFactory
import com.willfp.eco.internal.spigot.proxy.common.ai.target.NearestAttackableGoalFactory
import com.willfp.eco.internal.spigot.proxy.common.ai.target.NearestAttackableWitchGoalFactory
import com.willfp.eco.internal.spigot.proxy.common.ai.target.NearestHealableRaiderGoalFactory
import com.willfp.eco.internal.spigot.proxy.common.ai.target.NonTameRandomGoalFactory
import com.willfp.eco.internal.spigot.proxy.common.ai.target.OwnerHurtByGoalFactory
import com.willfp.eco.internal.spigot.proxy.common.ai.target.OwnerTargetGoalFactory
import com.willfp.eco.internal.spigot.proxy.common.getVersionSpecificEntityGoalFactory
import net.minecraft.world.entity.PathfinderMob
import net.minecraft.world.entity.ai.goal.Goal

fun <T : TargetGoal<*>> T.getGoalFactory(): TargetGoalFactory<T>? {
    val versionSpecific = this.getVersionSpecificEntityGoalFactory()
    if (versionSpecific != null) {
        return versionSpecific
    }

    @Suppress("UNCHECKED_CAST")
    return when (this) {
        is TargetGoalHurtBy -> HurtByGoalFactory
        is TargetGoalNearestAttackable -> NearestAttackableGoalFactory
        is TargetGoalNearestAttackableWitch -> NearestAttackableWitchGoalFactory
        is TargetGoalNearestHealableRaider -> NearestHealableRaiderGoalFactory
        is TargetGoalNonTameRandom -> NonTameRandomGoalFactory
        is TargetGoalOwnerHurtBy -> OwnerHurtByGoalFactory
        is TargetGoalOwnerTarget -> OwnerTargetGoalFactory
        is CustomGoal<*> -> CustomGoalFactory
        else -> null
    } as TargetGoalFactory<T>?
}

interface TargetGoalFactory<T : TargetGoal<*>> {
    fun create(apiGoal: T, entity: PathfinderMob): Goal?
    fun isGoalOfType(goal: Goal): Boolean
}
